File: chrome/android/java/src/org/chromium/chrome/browser/send_tab_to_self/SendTabToSelfAndroidBridge.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.send_tab_to_self;

import android.support.annotation.Nullable;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.profiles.Profile;

import java.util.ArrayList;
import java.util.List;

/**
 * Bridge to interface with send_tab_to_self_android_bridge which interacts with the corresponding
 * sync service. This is used by SendTabToSelfShareActivity when a user taps to share a tab. The
 * bridge is created and destroyed within the same method call.
 */
@JNINamespace("send_tab_to_self")
public class SendTabToSelfAndroidBridge {
    // TODO(tgupta): Add logic back in to track whether model is loaded.
    private boolean mIsNativeSendTabToSelfModelLoaded;

    /**
     * @param profile Profile of the user to retrieve the GUIDs for.
     * @returns All GUIDs for all SendTabToSelf entries
     */
    public static List<String> getAllGuids(Profile profile) {
        // TODO(tgupta): Add this assertion back in once the code to load is in place.
        // assert mIsNativeSendTabToSelfModelLoaded;
        List<String> toPopulate = new ArrayList<String>();
        SendTabToSelfAndroidBridgeJni.get().getAllGuids(profile, toPopulate);
        return toPopulate;
    }

    /**
     * Called by the native code in order to populate the list.
     *
     * @param allGuids List to populate provided by getAllGuids
     * @param newGuid The GUID to add to the list
     */
    @CalledByNative
    private static void addToGuidList(List<String> allGuids, String newGuid) {
        allGuids.add(newGuid);
    }

    /** Deletes all SendTabToSelf entries. This is called when the user disables sync. */
    public static void deleteAllEntries(Profile profile) {
        // TODO(tgupta): Add this assertion back in once the code to load is in place.
        // assert mIsNativeSendTabToSelfModelLoaded;
        SendTabToSelfAndroidBridgeJni.get().deleteAllEntries(profile);
    }

    /**
     * Creates a new entry to be persisted to the sync backend.
     *
     * @param profile Profile of the user to add entry for.
     * @param url URL to be shared
     * @param title Title of the page
     * @return The persisted entry which contains additional information such as the GUID or null in
     *     the case of an error.
     */
    @Nullable
    public static SendTabToSelfEntry addEntry(Profile profile, String url, String title) {
        // TODO(tgupta): Add this assertion back in once the code to load is in place.
        // assert mIsNativeSendTabToSelfModelLoaded;
        return SendTabToSelfAndroidBridgeJni.get().addEntry(profile, url, title);
    }

    /**
     * Return the entry associated with a particular GUID
     *
     * @param profile Profile of the user to get entry for.
     * @param guid The GUID to retrieve the entry for
     * @return The found entry or null if none exists
     */
    @Nullable
    public static SendTabToSelfEntry getEntryByGUID(Profile profile, String guid) {
        // TODO(tgupta): Add this assertion back in once the code to load is in place.
        // assert mIsNativeSendTabToSelfModelLoaded;
        return SendTabToSelfAndroidBridgeJni.get().getEntryByGUID(profile, guid);
    }

    /**
     * Deletes the entry associated with the GUID.
     *
     * @param profile Profile of the user to delete entry for.
     * @param guid The GUID to delete the entry for.
     */
    public static void deleteEntry(Profile profile, String guid) {
        SendTabToSelfAndroidBridgeJni.get().deleteEntry(profile, guid);
    }

    /**
     * Dismiss the entry associated with the GUID.
     *
     * @param profile Profile of the user to dismiss entry for.
     * @param guid The GUID to dismiss the entry for.
     */
    public static void dismissEntry(Profile profile, String guid) {
        SendTabToSelfAndroidBridgeJni.get().dismissEntry(profile, guid);
    }

    @NativeMethods
    interface Natives {
        SendTabToSelfEntry addEntry(Profile profile, String url, String title);

        void getAllGuids(Profile profile, List<String> guids);

        void deleteAllEntries(Profile profile);

        void deleteEntry(Profile profile, String guid);

        void dismissEntry(Profile profile, String guid);

        SendTabToSelfEntry getEntryByGUID(Profile profile, String guid);
    }
}

File: chrome/browser/android/send_tab_to_self/send_tab_to_self_android_bridge.cc
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <string>
#include <vector>

#include "base/android/jni_array.h"
#include "base/android/jni_string.h"
#include "base/metrics/histogram_macros.h"
#include "base/time/time.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_android.h"
#include "chrome/browser/send_tab_to_self/send_tab_to_self_client_service.h"
#include "chrome/browser/send_tab_to_self/send_tab_to_self_client_service_factory.h"
#include "chrome/browser/send_tab_to_self/send_tab_to_self_util.h"
#include "chrome/browser/sync/send_tab_to_self_sync_service_factory.h"
#include "components/send_tab_to_self/send_tab_to_self_entry.h"
#include "components/send_tab_to_self/send_tab_to_self_model.h"
#include "components/send_tab_to_self/send_tab_to_self_sync_service.h"
#include "content/public/browser/web_contents.h"
#include "jni/SendTabToSelfAndroidBridge_jni.h"
#include "jni/SendTabToSelfEntry_jni.h"
#include "url/gurl.h"

using base::android::AttachCurrentThread;
using base::android::ConvertUTF8ToJavaString;
using base::android::JavaParamRef;
using base::android::JavaRef;
using base::android::ScopedJavaGlobalRef;
using base::android::ScopedJavaLocalRef;

// The delegate to fetch SendTabToSelf information and persist new
// SendTabToSelf entries. The functions are called by the SendTabToSelf Java
// counterpart.
namespace send_tab_to_self {

namespace {

ScopedJavaLocalRef<jobject> CreateJavaSendTabToSelfEntry(
    JNIEnv* env,
    const SendTabToSelfEntry* entry) {
  return Java_SendTabToSelfEntry_createSendTabToSelfEntry(
      env, ConvertUTF8ToJavaString(env, entry->GetGUID()),
      ConvertUTF8ToJavaString(env, entry->GetURL().spec()),
      ConvertUTF8ToJavaString(env, entry->GetTitle()),
      entry->GetSharedTime().ToJavaTime(),
      entry->GetOriginalNavigationTime().ToJavaTime(),
      ConvertUTF8ToJavaString(env, entry->GetDeviceName()),
      ConvertUTF8ToJavaString(env, entry->GetTargetDeviceSyncCacheGuid()));
}

void LogModelLoadedInTime(bool status) {
  UMA_HISTOGRAM_BOOLEAN("SendTabToSelf.Sync.ModelLoadedInTime", status);
}

SendTabToSelfModel* GetModel(const JavaParamRef<jobject>& j_profile) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(j_profile);
  SendTabToSelfModel* model = SendTabToSelfSyncServiceFactory::GetInstance()
                                  ->GetForProfile(profile)
                                  ->GetSendTabToSelfModel();
  LogModelLoadedInTime(model->IsReady());
  return model;
}

}  // namespace

// Populates a list of GUIDs in the model.
static void JNI_SendTabToSelfAndroidBridge_GetAllGuids(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jobject>& j_guid_list_obj) {
  SendTabToSelfModel* model = GetModel(j_profile);
  if (model->IsReady()) {
    std::vector<std::string> all_ids = model->GetAllGuids();
    for (std::vector<std::string>::iterator it = all_ids.begin();
         it != all_ids.end(); ++it) {
      ScopedJavaLocalRef<jstring> j_guid = ConvertUTF8ToJavaString(env, *it);
      Java_SendTabToSelfAndroidBridge_addToGuidList(env, j_guid_list_obj,
                                                    j_guid);
    }
  }
}

// Deletes all entries in the model.
static void JNI_SendTabToSelfAndroidBridge_DeleteAllEntries(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile) {
  SendTabToSelfModel* model = GetModel(j_profile);
  if (model->IsReady()) {
    model->DeleteAllEntries();
  }
}

// Adds a new entry with the specified parameters. Returns the persisted
// version which contains additional information such as GUID.
static ScopedJavaLocalRef<jobject> JNI_SendTabToSelfAndroidBridge_AddEntry(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jstring>& j_url,
    const JavaParamRef<jstring>& j_title,
    jlong j_navigation_time,
    const JavaParamRef<jstring>& j_target_device_sync_cache_guid) {
  const std::string url = ConvertJavaStringToUTF8(env, j_url);
  const std::string title = ConvertJavaStringToUTF8(env, j_title);
  const std::string target_device_sync_cache_guid =
      ConvertJavaStringToUTF8(env, j_target_device_sync_cache_guid);
  base::Time navigation_time = base::Time::FromJavaTime(j_navigation_time);

  SendTabToSelfModel* model = GetModel(j_profile);
  if (!model->IsReady()) {
    return nullptr;
  }

  const SendTabToSelfEntry* persisted_entry = model->AddEntry(
      GURL(url), title, navigation_time, target_device_sync_cache_guid);

  if (persisted_entry == nullptr) {
    return nullptr;
  }

  return CreateJavaSendTabToSelfEntry(env, persisted_entry);
}

// Returns the entry associated with a GUID. May return nullptr if none is
// found.
static ScopedJavaLocalRef<jobject>
JNI_SendTabToSelfAndroidBridge_GetEntryByGUID(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jstring>& j_guid) {
  SendTabToSelfModel* model = GetModel(j_profile);
  if (!model->IsReady()) {
    return nullptr;
  }

  const std::string guid = ConvertJavaStringToUTF8(env, j_guid);
  const SendTabToSelfEntry* found_entry = model->GetEntryByGUID(guid);

  if (found_entry == nullptr) {
    return nullptr;
  }

  return CreateJavaSendTabToSelfEntry(env, found_entry);
}

// Deletes the entry associated with the passed in GUID.
static void JNI_SendTabToSelfAndroidBridge_DeleteEntry(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jstring>& j_guid) {
  SendTabToSelfModel* model = GetModel(j_profile);
  if (model->IsReady()) {
    const std::string guid = ConvertJavaStringToUTF8(env, j_guid);
    model->DeleteEntry(guid);
  }
}

// Marks the entry with the associated GUID as dismissed.
static void JNI_SendTabToSelfAndroidBridge_DismissEntry(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jstring>& j_guid) {
  SendTabToSelfModel* model = GetModel(j_profile);
  if (model->IsReady()) {
    const std::string guid = ConvertJavaStringToUTF8(env, j_guid);
    model->DismissEntry(guid);
  }
}

// Returns whether the feature is available for the specified |web_contents|.
static jboolean JNI_SendTabToSelfAndroidBridge_IsFeatureAvailable(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_web_contents) {
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(j_web_contents);

  return ShouldOfferFeature(web_contents);
}

}  // namespace send_tab_to_self
